Sem. 03/Pract. 03/mySolutions/Task02.cpp

```diff
@@ -0,0 +1,329 @@
+
+	std::cout << "Enter company name: ";
+	std::cin >> companyName;
+
+	if (existOffer(ALL_OFFERS_BINARY_FILE_PATH, companyName))
+	{
+		std::cout << "The company already exists!" << std::endl;
+		readJobOfferFromConsole(offer);
+	}
+	else
+	{
+		std::cout << "Enter Number of coworkers: ";
+		std::cin >> coworkers;
+		std::cout << "Enter number of days of paid leave: ";
+		std::cin >> numberOfVacancyDays;
+		std::cout << "Enter salary: ";
+		std::cin >> salary;
+
+		offer = createJobOffer(companyName, coworkers, numberOfVacancyDays, salary);
+	}
+}
+
+void printJobOffer(JobOffer& offer)
+{
+	std::cout << "Company Name: " << offer.companyName
+		<< "  Number of coworkers: " << offer.coworkers
+		<< "  Number of vacancy days: " << offer.numberOfVacancyDays
+		<< "  Salary: " << offer.salary
+		<< std::endl;
+}
+
+void filterBySalaryOffers(const char* filePath, const long long minSalary)
+{
+	std::ifstream in(filePath, std::ios::binary);
+
+	if (!in.is_open())
+	{
+		std::cerr << "Couldn't find file!";
+		return;
+	}
+
+	while (!in.eof())
+	{
+		JobOffer offer = readJobOfferFromFile(in);
+
+		if (offer.salary >= minSalary)
+			printJobOffer(offer);
+	}
+
+	in.close();
+}
+
+void readNJobOffersFromConsole()
+{
+	size_t numberOfOffers;
+	std::cout << "Enter number of job offers: ";
+	std::cin >> numberOfOffers;
+
+	JobOffer* offers = new JobOffer[numberOfOffers];
+
+	for (size_t i = 0; i < numberOfOffers; i++)
+		readJobOfferFromConsole(offers[i]);
+
+	saveJobOffers(offers, numberOfOffers);
+
+	delete[] offers;
+}
+
+bool isPerfectOffer(const JobOffer& offer, const int maxCoworkers, const int minVacancyDays, const long long minSalary)
+{
+	return offer.coworkers <= maxCoworkers && offer.numberOfVacancyDays >= minVacancyDays && offer.salary >= minSalary;
+}
+
+void perfectOffer(const char* filePath, const int maxCoworkers, const int minVacancyDays, const long long minSalary)
+{
+	std::ifstream in(ALL_OFFERS_BINARY_FILE_PATH, std::ios::binary);
+	std::ofstream out(PERFECT_OFFERS_BINARY_FILE_PATH, std::ios::app, std::ios::binary);
+
+	if (!in.is_open())
+	{
+		std::cerr << "Couldn't find file to read from!";
+		return;
+	}
+
+	if (!out.is_open())
+	{
+		std::cerr << "Couldn't find file to write into!";
+		return;
+	}
+
+	while (!in.eof())
+	{
+		JobOffer offer = readJobOfferFromFile(in);
+
+		if (isPerfectOffer(offer, maxCoworkers, minVacancyDays, minSalary) && !existOffer(PERFECT_OFFERS_BINARY_FILE_PATH, offer.companyName))
+			saveJobOffer(out, offer);
+	}
+
+	in.close();
+	out.close();
+}
+
+void printAllJobOffersFromFile(const char* filePath)
+{
+	std::ifstream in(filePath, std::ios::binary);
+
+	if (!in.is_open())
+	{
+		std::cerr << "Couldn't find fie!";
+		return;
+	}
+
+	size_t fileSize = getFileSize(in);
+	size_t numberOfOffers = fileSize / sizeof(JobOffer);
+	JobOffer* offers = new JobOffer[numberOfOffers];
+
+	in.read((char*)offers, fileSize);
+
+	for (size_t i = 0; i < numberOfOffers; i++)
+		printJobOffer(offers[i]);
+
+	in.close();
```

Comment:
close the stream when you don't need it anymore. in this case - on line 251

Reply:
Okay